worker.go

```diff
@@ -29,95 +56,183 @@ func rx(
+	}
+
+	// Get responseInterval from metadata, conditionally overriding it with the
+	// value loaded from the configuration file.
+	responseIntervalString, has := metadata["response_interval"]
+	if !has {
+		log.Warn("metadata missing response_interval, defaulting to 300")
+		responseIntervalString = "300"
+	}
+	responseInterval, err := time.ParseDuration(responseIntervalString + "s")
+	if err != nil {
+		return fmt.Errorf("cannot parse response interval: err=%w", err)
+	}
+	if config.DefaultConfig.ResponseInterval > 0 {
+		responseInterval = config.DefaultConfig.ResponseInterval
+	}
+
+	// Adjust responseInterval for batching mode.
+	if config.DefaultConfig.BatchEvents > 0 {
+		// Set the response interval to 500ms when batching events. This has the
+		// effect of matching the "<-timeout" case every time the channel select
+		// statement evaluates. This allows the same codepath to work when
+		// either batching events by quantity or by timeout.
+		responseInterval = 500 * time.Millisecond
 	}
 
 	if config.DefaultConfig.VerifyPlaybook {
 		d, err := verifyPlaybook(data, config.DefaultConfig.InsightsCoreGPGCheck)
 		if err != nil {
-			return fmt.Errorf("cannot verify playbook: %v", err)
+			return fmt.Errorf("cannot verify playbook: err=%w", err)
 		}
 		data = d
 	}
 
-	events, err := ansible.RunPlaybook(id, data, correlationID)
+	// Create the event manager.
+	eventManager := NewEventManager(id, returnURL, responseInterval, w)
+
+	// Create the playbook runner.
+	runner := ansible.NewRunner(correlationID, 60*time.Second)
+
+	// Start the goroutine processing events from the runner.
+	go eventManager.processEvents(runner)
+	go eventManager.transmitCachedEvents()
+
+	// Run the playbook.
+	err = runner.Run(data)
 	if err != nil {
-		return fmt.Errorf("cannot run playbook: %v", err)
+		return fmt.Errorf("cannot run playbook: err=%w", err)
 	}
 
-	// start a goroutine that receives ansible-runner events as they are
-	// emitted.
-	go func() {
-		// TODO: support metadata["response_interval"] batch processing
-		var cachedEvents []byte
-		for event := range events {
-			log.Debugf("ansible-runner event: %s", event)
+	return nil
+}
 
-			cachedEvents = append(cachedEvents, append(event, '\n')...)
+// processEvents receives values from the runner and caches them for future use.
+func (e *EventManager) processEvents(runner *ansible.Runner) {
+	for event := range runner.Events {
+		e.cachedEventsLock.Lock()
+		e.cachedEvents = append(e.cachedEvents, event)
+		e.cachedEventsLock.Unlock()
+	}
 
-			var runnerEvent map[string]interface{}
-			if err := json.Unmarshal(event, &runnerEvent); err != nil {
-				log.Errorf("cannot unmarshal JSON: %v", err)
-				continue
+	// Signal the sending events goroutine to stop.
+	e.stopSendingEvents <- struct{}{}
+
+	// Transmit one final batch of all events.
+	e.cachedEventsLock.RLock()
+	length := len(e.cachedEvents)
+	e.cachedEventsLock.RUnlock()
+	if err := e.transmitEvents(0, length); err != nil {
```

Comment:
Calling `e.transmitEvents()` from two go routines (this and `transmitCachedEvents()`) could be risky, when `e.cachedEventLock` is unlocked. I think that it could lead to race conditions. The `e.transmitEvents()` could be called twice with same values or overlapping ranges.